File: src/backend/Mille/Mille.Domain/Entities/Order.cs
using Mille.Domain.Enums;
using Mille.Domain.Exceptions;
using System;
using System.Collections.Generic;
using System.Text;

namespace Mille.Domain.Entities
{
    public class Order
    {
        public Guid Id { get; private set; }
        public Guid UserId { get; private set; }
        public string ReceiverName { get; private set; }
        public string ReceiverPhone { get; private set; }
        public string ShippingAddress { get; private set; }
        public decimal TotalAmount { get; private set; }
        public decimal DiscountAmount { get; private set; }
        public OrderStatus Status { get; private set; }
        public DateTime CreatedAt { get; private set; }
        public DateTime UpdatedAt { get; private set; }

        public User User { get; private set; }
        public Payment Payment { get; private set; }

        private readonly List<OrderItem> _items = new();
        public IReadOnlyCollection<OrderItem> Items => _items.AsReadOnly();

        private readonly List<OrderStatusHistory> _statusHistories = new();
        public IReadOnlyCollection<OrderStatusHistory> StatusHistories => _statusHistories.AsReadOnly();

        private Order() { }

        public Order(Guid userId, string receiverName, string receiverPhone, string shippingAddress, decimal totalAmount, decimal discountAmount = 0)
        {
            Id = Guid.NewGuid();
            UserId = userId;
            ReceiverName = receiverName;
            ReceiverPhone = receiverPhone;
            ShippingAddress = shippingAddress;
            TotalAmount = totalAmount;
            DiscountAmount = discountAmount;
            Status = OrderStatus.Pending;
            CreatedAt = DateTime.UtcNow;
            UpdatedAt = DateTime.UtcNow;

            AddStatusHistory(OrderStatus.Pending, "Đơn hàng đã được khởi tạo.");
        }

        public OrderItem AddItem(Guid productVariantId, string productName, string sku, int quantity, decimal unitPrice)
        {
            if (Status != OrderStatus.Pending)
                throw new DomainException("Chỉ có thể thêm sản phẩm khi đơn hàng ở trạng thái Chờ xử lý.");

            var item = new OrderItem(Id, productVariantId, productName, sku, quantity, unitPrice);
            _items.Add(item);
            UpdatedAt = DateTime.UtcNow;
            return item;
        }

        public void AttachPayment(Payment payment)
        {
            Payment = payment;
            UpdatedAt = DateTime.UtcNow;
        }

        public void Confirm(string? note = null)
        {
            ChangeStatus(OrderStatus.Confirmed, note ?? "Đơn hàng đã được xác nhận.");
        }

        public void Ship(string? note = null)
        {
            if (Status != OrderStatus.Confirmed)
                throw new DomainException("Chỉ có thể giao đơn hàng đã được xác nhận.");

            ChangeStatus(OrderStatus.Shipping, note ?? "Đơn hàng đang được giao.");
        }

        public void Complete(string? note = null)
        {
            if (Status != OrderStatus.Shipping  )
                throw new DomainException("Chỉ có thể hoàn tất đơn hàng đang giao.");

            ChangeStatus(OrderStatus.Completed, note ?? "Giao hàng thành công. Đơn hàng hoàn tất.");
        }

        public void Cancel(string reason)
        {
            if (Status == OrderStatus.Completed || Status == OrderStatus.Cancelled)
                throw new DomainException("Không thể hủy đơn hàng đã hoàn tất hoặc đã hủy trước đó.");

            ChangeStatus(OrderStatus.Cancelled, $"Đơn hàng bị hủy. Lý do: {reason}");
        }

        private void AddStatusHistory(OrderStatus status, string? note = null)
        {
            var newstatusHistory = new OrderStatusHistory(Id, status, note);
            _statusHistories.Add(newstatusHistory);
        }

        private void ChangeStatus(OrderStatus newStatus, string note)
        {
            Status = newStatus;
            UpdatedAt = DateTime.UtcNow;
            _statusHistories.Add(new OrderStatusHistory(Id, newStatus, note));
        }
    }
}

File: src/frontend/mille-ecommerce/src/modules/orders/types/order.ts
import { z } from 'zod'

export type OrderStatus = 'Pending' | 'Confirmed' | 'Shipping' | 'Completed' | 'Cancelled'

export const ORDER_STATUS_LABELS: Record<OrderStatus, string> = {
  Pending: 'Pending',
  Confirmed: 'Confirmed',
  Shipping: 'Shipping',
  Completed: 'Completed',
  Cancelled: 'Cancelled',
}

export type PaymentMethod = 'COD' | 'VNPay' | 'Stripe'

export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  COD: 'COD',
  VNPay: 'VNPay',
  Stripe: 'Stripe',
}

export type PaymentStatus = 'Pending' | 'Success' | 'Failed' | 'Refunded'

export const PAYMENT_STATUS_LABELS: Record<PaymentStatus, string> = {
  Pending: 'Pending',
  Success: 'Success',
  Failed: 'Failed',
  Refunded: 'Refunded',
}

export const createOrderSchema = z.object({
  receiverName: z
    .string()
    .nonempty('Receiver name is required')
    .max(100, 'Receiver name must not exceed 100 characters'),
  receiverPhone: z
    .string()
    .nonempty('Receiver phone is required')
    .max(20, 'Receiver phone must not exceed 20 characters'),
  shippingAddress: z
    .string()
    .nonempty('Shipping address is required')
    .max(500, 'Shipping address must not exceed 500 characters'),
  paymentMethod: z.enum(['COD', 'VNPay', 'Stripe'], {
    message: 'Invalid payment method',
  }),
})

export const updateOrderStatusSchema = z.object({
  newStatus: z.enum(['Pending', 'Confirmed', 'Shipping', 'Completed', 'Cancelled'], {
    message: 'Invalid status',
  }),
  note: z.string().max(500, 'Note must not exceed 500 characters').optional(),
})

export type CreateOrderRequest = z.infer<typeof createOrderSchema>
export type UpdateOrderStatusRequest = z.infer<typeof updateOrderStatusSchema>

export type CreateOrderResponse = {
  orderId: string
  totalAmount: number
  paymentMethod: PaymentMethod
}

export type UpdateOrderStatusResponse = {
  message: string
}

export type OrdersResponse = {
  orders: OrderSummaryDto[]
  totalCount: number
  page: number
  pageSize: number
  totalPages: number
}

export type GetOrderResponse = {
  id: string
  receiverName: string
  receiverPhone: string
  shippingAddress: string
  totalAmount: number
  discountAmount: number
  status: string
  createdAt: string
  updatedAt: string
  items: OrderItemDto[]
  statusHistories: OrderStatusHistoryDto[]
  payment: PaymentDto | null
}

export type CancelOrderResponse = {
  message: string
}

export type ConfirmPaymentResponse = {
  status: string
  paidAt: string | null
  message: string
}

export type OrderSummaryDto = {
  id: string
  receiverName: string
  receiverPhone: string
  shippingAddress: string
  totalAmount: number
  status: string
  paymentMethod: string
  paymentStatus: string
  createdAt: string
}

export type OrderItemDto = {
  id: number
  productVariantId: string
  productName: string
  sku: string
  quantity: number
  unitPrice: number
  subtotal: number
}

export type OrderStatusHistoryDto = {
  id: number
  status: string
  note: string | null
  createdAt: string
}

export type PaymentDto = {
  id: string
  method: string
  amount: number
  transactionId: string | null
  gatewayResponse: string | null
  status: string
  paidAt: string | null
  createdAt: string
}
